File: compat-content/src/main/java/com/hendraanggrian/compat/content/MimeTypes.java
package com.hendraanggrian.compat.content;

import android.content.ContentResolver;
import android.content.Context;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.webkit.MimeTypeMap;

import java.io.File;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public final class MimeTypes {

    private MimeTypes() {
    }

    @Nullable
    public static String guessUri(@NonNull Context context, @NonNull Uri uri) {
        return uri.getScheme().equals(ContentResolver.SCHEME_CONTENT)
                ? context.getContentResolver().getType(uri)
                : guessURL(Uri.fromFile(new File(uri.getPath())).toString());
    }

    @Nullable
    public static String guessURL(@Nullable String url) {
        return guessExtension(MimeTypeMap.getFileExtensionFromUrl(url));
    }

    @Nullable
    public static String guessExtension(@Nullable String extension) {
        return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
    }
}

File: kota/src/kota/MimeTypes.kt
@file:Suppress("NOTHING_TO_INLINE", "UNUSED")

package kota

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.webkit.MimeTypeMap.getFileExtensionFromUrl
import android.webkit.MimeTypeMap.getSingleton
import java.io.File

inline val String.mimeType: String get() = getFileExtensionFromUrl(this)

inline val String.mimeExtension: String get() = getSingleton().getExtensionFromMimeType(this)

inline fun Uri.getMimeType(resolver: ContentResolver): String = if (isContent) resolver.getType(this) else File(path).absolutePath.mimeType
inline fun Uri.getMimeType(context: Context): String = getMimeType(context.contentResolver)